### ISSUE TITLE
Ability to extend CompileCte in child compilers

### ISSUE THREAD

**Issue by slimbr** [sqlkata/querybuilder#230]
Hi.
i want use cte query
`var queryParent = new Query("tbl").Select("tblid").Where("objid", objId);`
`var queryChild = new Query("tbl").Select("tbl.tblid").Join("rem", "rem.tblid", "tbl.parentid");`
`var cte = new Query().With("rem", queryParent.UnionAll(queryChild));`
`var queryCte = new Query("rem").Select("tblid");`
`var queryDelete = cte.From("tbltbl").AsDelete().WhereIn("tblid", queryCte);`
MsSql - success, use "WITH"
Postgres - fail, need use "WITH RECURSIVE" instead "WITH"

File SqlKata.Compilers.Compiler method CompileRaw line 122

**Comment by slimbr** [comment]
resolve #231

**Comment by ahmad-moussawi** [comment]
Sorry but It's not clear what is exactly your issue, providing the expected SQL string could help.

btw I think your query can be written like so:

```cs
var unionQuery = new Query("table").Select("id").Where("SomeId", id)
    .Union(new Query("table").Select("id").Join("..."));
var deleteQuery = new Query("table2").WhereIn("id", unionQuery).AsDelete();
```

**Comment by slimbr** [comment]
the fact is that Postgres sql for hierarchical queries requires a command
"WITH RECURSIVE"
MsSql requires a command
"WITH"

source code Compiler.cs method CompileRaw contains instruction as a string "WITH".
in order to redefine only cte build it is necessary to bring the code to the virtual method.

sorry, but proposed query will not work.

sample test mssql:

`begin transaction`
`create table tbl`
`(`
`	tblid int,`
`	parentid int,`
`	objid nvarchar(max)`
`)`
`insert into tbl`
`select 1, null, 'A'`
`insert into tbl`
`select 2, 1, 'B'`
`insert into tbl`
`select 3, 2, 'C'`
`insert into tbl`
`select 4, 3, 'D'`
`;WITH rem AS`
`(`
`SELECT tblid`
`FROM tbl`
`WHERE objid = 'B'`
`UNION ALL`
`SELECT tbl.tblid FROM tbl`
`INNER JOIN rem ON rem.tblid = tbl.parentid`
`)`
`select * from rem`
`rollback transaction`

**Comment by for7raid** [comment]
Hi Ahmad!
> btw I think your query can be written like so:

unfortunate no, with code is not the same. In our case we have to create a tree recursive subselect, and then join the temporary table result 'cte' in next queries. so this is not a simple union.

'recursive' word is necessary in PostgreSQL in this case.

The code I expected after compile is:

```sql
with recursive rem as (
select tblid from tbl
union all
select tblid from tbl join rem on tbl.parentid = rem.tblid 
)
delete from tbltbl where tbltblid in (select tblid from rem); --while pgsql not support delete-join operations
```

**Comment by ahmad-moussawi** [comment]
Thanks, guys for the info, I merged the PR #231

**Comment by for7raid** [comment]
thanks a lot!
but anyway, please think  about original issue, i suggest to change compilecte code in pgcompiler.
